File: src/commands/debug.ts
import Anthropic from '@anthropic-ai/sdk';
import type { Intensity } from '../types.js';
import { getClient, MissingApiKeyError } from '../llm/client.js';
import { createSession, finalizeSession } from '../core/session.js';
import { runSessionLoop } from '../core/session-loop.js';
import { saveSession } from '../storage/session-store.js';
import { tagSessionBehaviors } from '../scoring/behavior-tagger.js';
import { computeSkillScores } from '../scoring/skill-scorer.js';
import { computeGrowthProfile } from '../scoring/growth-profile.js';
import { selectClosingQuote } from '../quotes/selector.js';
import { renderGrowthScore } from '../ui/growth-display.js';
import { displayError, displayApiKeyHelp } from '../ui/renderer.js';

export async function debugCommand(
  problemDescription: string,
  intensity: Intensity,
): Promise<void> {
  // Validate API key
  try {
    getClient();
  } catch (error) {
    if (error instanceof MissingApiKeyError) {
      displayApiKeyHelp();
      process.exit(1);
    }
    throw error;
  }

  // Validate problem description
  if (!problemDescription || !problemDescription.trim()) {
    displayError('Please provide a problem description.');
    console.log('\nUsage: nietzschian debug "Your problem description here"');
    process.exit(2);
  }

  // Create session
  const session = createSession(problemDescription.trim(), intensity);

  try {
    // Run interactive session
    const result = await runSessionLoop(session);

    // Finalize session
    let behaviorTags = session.behaviorTags;
    let skillScores = session.skillScores;

    try {
      behaviorTags = await tagSessionBehaviors(session.transcript);
      skillScores = computeSkillScores(behaviorTags);
    } catch {
      // Scoring is best-effort — don't crash if it fails
    }

    finalizeSession(session, result.outcome, skillScores, behaviorTags);

    // Save session
    try {
      await saveSession(session);
    } catch {
      // Persistence is best-effort
    }

    // Display growth score
    const growthProfile = await computeGrowthProfile();
    const closingQuote = selectClosingQuote(result.outcome);
    const display = renderGrowthScore(session, growthProfile, closingQuote);
    console.log(display);

    process.exit(0);
  } catch (error) {
    if (error instanceof Anthropic.AuthenticationError) {
      displayError('Invalid API key. Check your ANTHROPIC_API_KEY.');
      process.exit(1);
    }
    if (error instanceof Anthropic.RateLimitError) {
      displayError('Rate limit reached. Try again in a moment.');
      process.exit(3);
    }
    if (error instanceof Anthropic.APIConnectionError) {
      displayError('Cannot reach Claude API. Check your connection.');
      process.exit(3);
    }
    if (error instanceof Anthropic.APIError) {
      displayError(`API error: ${error.message}`);
      process.exit(3);
    }

    throw error;
  }
}

File: nietzschian_debugger/commands/debug.py
"""Debug command - starts an interactive debugging session."""

from __future__ import annotations

import asyncio
import sys

import anthropic

from ..types import Intensity
from ..llm.client import get_client, MissingApiKeyError
from ..core.session import create_session, finalize_session
from ..core.session_loop import run_session_loop
from ..storage.session_store import save_session
from ..scoring.behavior_tagger import tag_session_behaviors
from ..scoring.skill_scorer import compute_skill_scores
from ..scoring.growth_profile import compute_growth_profile
from ..quotes.selector import select_closing_quote
from ..ui.growth_display import render_growth_score
from ..ui.renderer import display_error, display_api_key_help


async def debug_command(problem_description: str, intensity: Intensity) -> None:
    """Run a full debugging session."""
    # Validate API key
    try:
        get_client()
    except MissingApiKeyError:
        display_api_key_help()
        sys.exit(1)

    # Validate problem description
    if not problem_description or not problem_description.strip():
        display_error("Please provide a problem description.")
        print('\nUsage: nietzschian debug "Your problem description here"')
        sys.exit(2)

    # Create session
    session = create_session(problem_description.strip(), intensity)

    try:
        # Run interactive session
        result = await run_session_loop(session)

        # Finalize session
        behavior_tags = session.behavior_tags
        skill_scores = session.skill_scores

        try:
            behavior_tags = await tag_session_behaviors(session.transcript)
            skill_scores = compute_skill_scores(behavior_tags)
        except Exception:
            # Scoring is best-effort
            pass

        finalize_session(session, result.outcome, skill_scores, behavior_tags)

        # Save session
        try:
            await save_session(session)
        except Exception:
            # Persistence is best-effort
            pass

        # Display growth score
        growth_profile = await compute_growth_profile()
        closing_quote = select_closing_quote(result.outcome)
        display = render_growth_score(session, growth_profile, closing_quote)
        print(display)

        sys.exit(0)

    except anthropic.AuthenticationError:
        display_error("Invalid API key. Check your ANTHROPIC_API_KEY.")
        sys.exit(1)
    except anthropic.RateLimitError:
        display_error("Rate limit reached. Try again in a moment.")
        sys.exit(3)
    except anthropic.APIConnectionError:
        display_error("Cannot reach Claude API. Check your connection.")
        sys.exit(3)
    except anthropic.APIError as e:
        display_error(f"API error: {e.message}")
        sys.exit(3)
